driver-core/src/main/java/com/datastax/driver/core/Frame.java

```diff
@@ -328,4 +332,16 @@ protected void encode(ChannelHandlerContext ctx, Frame frame, List<Object> out)
             }
         }
     }
+
+    @ChannelHandler.Sharable
+    static class BetaVersionFlagAdder extends MessageToMessageEncoder<Frame> {
+        @Override
+        protected void encode(ChannelHandlerContext ctx, Frame frame, List<Object> out) throws Exception {
+            ProtocolVersion protocolVersion = frame.header.version;
+            // Check for the protocol version, if the downgrade has occurred during renegotiation
+            if (protocolVersion.toInt() >= ProtocolVersion.V5.toInt())
+                frame.header.flags.add(Header.Flag.USE_BETA);
```

Comment:
> our driver would most likely fail to connect if USE_BETA was not requested (case B1)

I'm not sure how the current version of the PR behaves, but what I would expect is "connect with X-1" (the highest stable supported by both sides).

> It would probably also fail to connect even if the flag was requested (case B2), but that's more acceptable

Agreed. That also applies to the case where the server is on a _different_ X-beta -- i.e. more beta features were added to X since the driver started supporting it (see my [comment on 10786](https://issues.apache.org/jira/browse/CASSANDRA-10786?focusedCommentId=15344506&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15344506)).

> Request could convey the request for the flag, pretty much like isTracingRequested()

Yes, that would avoid the additional handler.

> If the user does not specify a version and do not call allowBetaVersions(), use Protocol.NEWEST_SUPPORTED

Wouldn't it be slightly simpler if we didn't allow auto-negotiation with beta versions?
If you want to use a beta, you would have to specify the version as well (we could even have an alternate method `Cluster.Builder#withBetaProtocolVersion(ProtocolVersion)`).
With auto-negotiation, you would always end up with a stable.